File: cpp/src/segmentation.cpp
#include "segmentation.hpp"

#include <algorithm>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <numeric>
#include <string>
#include <vector>

namespace {

constexpr int MIN_COMPONENT_AREA = 8;
constexpr int MIN_COMPONENT_WIDTH = 2;
constexpr int MIN_COMPONENT_HEIGHT = 2;
constexpr int BORDER_CLEAN_WIDTH = 8;
constexpr int TRACE_GAP_LIMIT_UPPER = 25;
constexpr int TRACE_GAP_LIMIT_LOWER = 50;
constexpr int MAX_EDGE_ARTIFACT_WIDTH = 24;
constexpr double MIN_EDGE_ARTIFACT_HEIGHT_FRACTION = 0.45;
constexpr int MAX_JUMP_UPPER = 80;
constexpr int MAX_JUMP_LOWER = 100;
constexpr int UPPER_MAX_INTERPOLATION_GAP = 20;
constexpr int LOWER_MAX_INTERPOLATION_GAP = 45;
constexpr double LOWER_SIGNAL_Y_END_FRACTION = 0.93;
constexpr double LOWER_SIGNAL_Y_END_FRACTION_STRICT = 0.86;
constexpr double LOWER_LONG_COMPONENT_Y_FRACTION = 0.70;
constexpr double LOWER_LONG_COMPONENT_ASPECT_RATIO = 12.0;
constexpr double LOWER_LONG_COMPONENT_WIDTH_FRACTION = 0.10;
constexpr int LOWER_LONG_COMPONENT_MAX_HEIGHT = 6;
constexpr double LOWER_RIGHT_TEXT_X_FRACTION = 0.60;
constexpr double LOWER_RIGHT_TEXT_Y_FRACTION = 0.55;
constexpr int LOWER_RIGHT_TEXT_MAX_AREA = 1200;
constexpr int LOWER_RIGHT_TEXT_MAX_HEIGHT = 90;
constexpr int LOWER_RIGHT_TEXT_MAX_WIDTH = 300;
constexpr int LOWER_HORIZONTAL_KERNEL_WIDTH = 40;
constexpr int LOWER_HORIZONTAL_MIN_WIDTH = 100;
constexpr double LOWER_HORIZONTAL_MIN_ASPECT_RATIO = 12.0;
constexpr int LOWER_HORIZONTAL_MAX_HEIGHT = 6;
constexpr int LOWER_BASELINE_KEEP_MIN_HEIGHT = 10;
constexpr int LOWER_BASELINE_KEEP_MIN_AREA = 30;
constexpr double LOWER_BASELINE_Y_FRACTION = 0.70;

void ensureParentDir(const std::string& path) {
    const std::filesystem::path parent = std::filesystem::path(path).parent_path();
    if (!parent.empty()) {
        std::filesystem::create_directories(parent);
    }
}

double getLowerRoiFraction(const std::string& cleanupMode) {
    if (cleanupMode == "strict") {
        return LOWER_SIGNAL_Y_END_FRACTION_STRICT;
    }
    return LOWER_SIGNAL_Y_END_FRACTION;
}

bool shouldRemoveUpperComponent(int maskHeight, int y, int area) {
    // Удаляем подписи и короткий шум сверху/снизу верхней панели.
    if (area < 15) {
        return true;
    }
    if (y < maskHeight * 0.08 && area < 800) {
        return true;
    }
    if (y > maskHeight * 0.92 && area < 800) {
        return true;
    }
    return false;
}

bool shouldRemoveLowerComponent(
    int maskWidth,
    int maskHeight,
    int x,
    int y,
    int width,
    int height,
    int area,
    const std::string& cleanupMode
) {
    const double aspectRatio = static_cast<double>(width) / std::max(1, height);

    if (area < 20) {
        return true;
    }

    const bool isLowerLongLine =
        y > maskHeight * LOWER_LONG_COMPONENT_Y_FRACTION &&
        aspectRatio > LOWER_LONG_COMPONENT_ASPECT_RATIO &&
        height <= LOWER_LONG_COMPONENT_MAX_HEIGHT &&
        width > maskWidth * LOWER_LONG_COMPONENT_WIDTH_FRACTION;

    const bool isRightLowerText =
        x > maskWidth * LOWER_RIGHT_TEXT_X_FRACTION &&
        y > maskHeight * LOWER_RIGHT_TEXT_Y_FRACTION &&
        area < LOWER_RIGHT_TEXT_MAX_AREA &&
        height < LOWER_RIGHT_TEXT_MAX_HEIGHT &&
        width < LOWER_RIGHT_TEXT_MAX_WIDTH;

    if (isLowerLongLine || isRightLowerText) {
        return true;
    }

    const bool isLowSignalLike =
        y > maskHeight * LOWER_BASELINE_Y_FRACTION &&
        height >= LOWER_BASELINE_KEEP_MIN_HEIGHT &&
        area >= LOWER_BASELINE_KEEP_MIN_AREA;

    if (cleanupMode == "soft" && isLowSignalLike) {
        return false;
    }

    return false;
}

bool shouldRemoveComponent(
    const std::string& panelType,
    int maskWidth,
    int maskHeight,
    int x,
    int y,
    int width,
    int height,
    int area,
    const std::string& cleanupMode
) {
    const double aspectRatio = static_cast<double>(width) / std::max(1, height);

    if (area < MIN_COMPONENT_AREA) {
        return true;
    }
    if (width < MIN_COMPONENT_WIDTH || height < MIN_COMPONENT_HEIGHT) {
        return true;
    }
    if (area < 40 && width < 10 && height < 10) {
        return true;
    }
    if (aspectRatio > 15.0 && height < 5) {
        return true;
    }
    if (
        x + width >= maskWidth - BORDER_CLEAN_WIDTH &&
        width <= MAX_EDGE_ARTIFACT_WIDTH &&
        height >= maskHeight * MIN_EDGE_ARTIFACT_HEIGHT_FRACTION
    ) {
        return true;
    }

    if (panelType == "upper") {
        return shouldRemoveUpperComponent(maskHeight, y, area);
    }

    return shouldRemoveLowerComponent(
        maskWidth,
        maskHeight,
        x,
        y,
        width,
        height,
        area,
        cleanupMode
    );
}

cv::Mat filterSignalComponents(
    const cv::Mat& mask,
    const std::string& panelType,
    const std::string& cleanupMode,
    cv::Mat* removedOut = nullptr
) {
    cv::Mat labels;
    cv::Mat stats;
    cv::Mat centroids;
    const int componentCount = cv::connectedComponentsWithStats(mask, labels, stats, centroids, 8);

    cv::Mat filtered = cv::Mat::zeros(mask.size(), mask.type());
    cv::Mat removed = cv::Mat::zeros(mask.size(), mask.type());
    const int maskHeight = mask.rows;
    const int maskWidth = mask.cols;

    for (int componentId = 1; componentId < componentCount; ++componentId) {
        const int x = stats.at<int>(componentId, cv::CC_STAT_LEFT);
        const int y = stats.at<int>(componentId, cv::CC_STAT_TOP);
        const int width = stats.at<int>(componentId, cv::CC_STAT_WIDTH);
        const int height = stats.at<int>(componentId, cv::CC_STAT_HEIGHT);
        const int area = stats.at<int>(componentId, cv::CC_STAT_AREA);

        cv::Mat componentMask = labels == componentId;
        if (shouldRemoveComponent(panelType, maskWidth, maskHeight, x, y, width, height, area, cleanupMode)) {
            removed.setTo(255, componentMask);
        } else {
            filtered.setTo(255, componentMask);
        }
    }

    if (removedOut != nullptr) {
        *removedOut = removed;
    }
    return filtered;
}

cv::Mat removeLowerHorizontalArtifacts(
    const cv::Mat& mask,
    const std::string& panelType,
    const std::string& cleanupMode,
    cv::Mat* removedOut = nullptr
) {
    (void)cleanupMode;

    cv::Mat removed = cv::Mat::zeros(mask.size(), mask.type());
    if (panelType != "lower") {
        if (removedOut != nullptr) {
            *removedOut = removed;
        }
        return mask.clone();
    }

    cv::Mat horizontalKernel = cv::getStructuringElement(
        cv::MORPH_RECT,
        cv::Size(LOWER_HORIZONTAL_KERNEL_WIDTH, 3)
    );
    cv::Mat horizontalMask;
    cv::morphologyEx(mask, horizontalMask, cv::MORPH_OPEN, horizontalKernel);
    horizontalMask.rowRange(0, static_cast<int>(mask.rows * LOWER_LONG_COMPONENT_Y_FRACTION)).setTo(0);

    cv::Mat labels;
    cv::Mat stats;
    cv::Mat centroids;
    const int componentCount = cv::connectedComponentsWithStats(horizontalMask, labels, stats, centroids, 8);

    for (int componentId = 1; componentId < componentCount; ++componentId) {
        const int width = stats.at<int>(componentId, cv::CC_STAT_WIDTH);
        const int height = stats.at<int>(componentId, cv::CC_STAT_HEIGHT);
        const double aspectRatio = static_cast<double>(width) / std::max(1, height);

        if (
            width >= LOWER_HORIZONTAL_MIN_WIDTH &&
            aspectRatio >= LOWER_HORIZONTAL_MIN_ASPECT_RATIO &&
            height <= LOWER_HORIZONTAL_MAX_HEIGHT
        ) {
            removed.setTo(255, labels == componentId);
        }
    }

    cv::Mat cleaned = mask.clone();
    cleaned.setTo(0, removed > 0);

    if (removedOut != nullptr) {
        *removedOut = removed;
    }
    return cleaned;
}

int medianY(const std::vector<int>& values) {
    if (values.empty()) {
        return 0;
    }

    std::vector<int> sorted = values;
    std::sort(sorted.begin(), sorted.end());
    const size_t middle = sorted.size() / 2;
    if (sorted.size() % 2 == 1) {
        return sorted[middle];
    }
    return static_cast<int>(std::llround((sorted[middle - 1] + sorted[middle]) / 2.0));
}

void drawLabel(cv::Mat& image, int x, int y, const std::string& text, const cv::Scalar& color) {
    const int font = cv::FONT_HERSHEY_SIMPLEX;
    const double scale = 0.65;
    const int thickness = 2;
    int baseline = 0;
    const cv::Size textSize = cv::getTextSize(text, font, scale, thickness, &baseline);

    x = std::max(0, std::min(image.cols - textSize.width - 12, x));
    y = std::max(textSize.height + 8, std::min(image.rows - 8, y));

    cv::rectangle(
        image,
        cv::Point(x - 4, y - textSize.height - 8),
        cv::Point(x + textSize.width + 8, y + 6),
        cv::Scalar(255, 255, 255),
        cv::FILLED
    );
    cv::putText(image, text, cv::Point(x, y), font, scale, color, thickness, cv::LINE_AA);
}

}  // namespace

PanelSplit splitPanels(const cv::Mat& alignedImage) {
    const int height = alignedImage.rows;

    const int upperStart = static_cast<int>(height * 0.05);
    const int upperEnd = static_cast<int>(height * 0.60);
    const int lowerStart = static_cast<int>(height * 0.62);
    const int lowerEnd = static_cast<int>(height * 0.98);

    PanelSplit split;
    split.upper = alignedImage.rowRange(upperStart, upperEnd).clone();
    split.lower = alignedImage.rowRange(lowerStart, lowerEnd).clone();
    return split;
}

cv::Mat extractDarkMask(const cv::Mat& panelImage) {
    cv::Mat gray;
    cv::Mat blurred;
    cv::Mat rawMask;

    cv::cvtColor(panelImage, gray, cv::COLOR_BGR2GRAY);
    cv::GaussianBlur(gray, blurred, cv::Size(5, 5), 0);
    cv::threshold(blurred, rawMask, 160, 255, cv::THRESH_BINARY_INV);
    return rawMask;
}

cv::Mat cleanSignalMask(
    const cv::Mat& mask,
    const std::string& panelType,
    const std::string& cleanupMode
) {
    cv::Mat clean = mask.clone();

    // Убираем темные полосы по краям фотографии.
    clean.colRange(0, std::min(BORDER_CLEAN_WIDTH, clean.cols)).setTo(0);
    clean.colRange(std::max(0, clean.cols - BORDER_CLEAN_WIDTH), clean.cols).setTo(0);
    clean.rowRange(0, std::min(BORDER_CLEAN_WIDTH, clean.rows)).setTo(0);
    clean.rowRange(std::max(0, clean.rows - BORDER_CLEAN_WIDTH), clean.rows).setTo(0);

    cv::Mat openKernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2, 2));
    cv::morphologyEx(clean, clean, cv::MORPH_OPEN, openKernel);

    if (panelType == "lower") {
        const int roiYEnd = static_cast<int>(clean.rows * getLowerRoiFraction(cleanupMode));
        clean.rowRange(roiYEnd, clean.rows).setTo(0);
    }

    cv::Mat filtered = filterSignalComponents(clean, panelType, cleanupMode);

    cv::Mat closeKernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2, 3));
    cv::morphologyEx(filtered, filtered, cv::MORPH_CLOSE, closeKernel);

    filtered = removeLowerHorizontalArtifacts(filtered, panelType, cleanupMode);
    filtered = filterSignalComponents(filtered, panelType, cleanupMode);

    return filtered;
}

std::vector<cv::Point> extractTracePoints(const cv::Mat& mask, const std::string& panelType) {
    std::vector<cv::Point> points;
    int previousY = -1;
    const int maxJump = panelType == "upper" ? MAX_JUMP_UPPER : MAX_JUMP_LOWER;

    for (int x = 0; x < mask.cols; ++x) {
        std::vector<int> ys;
        ys.reserve(mask.rows / 4);

        for (int y = 0; y < mask.rows; ++y) {
            if (mask.at<uchar>(y, x) > 0) {
                ys.push_back(y);
            }
        }

        if (ys.empty()) {
            previousY = -1;
            continue;
        }

        int y = 0;
        if (previousY < 0) {
            y = medianY(ys);
        } else {
            auto closest = std::min_element(
                ys.begin(),
                ys.end(),
                [previousY](int a, int b) {
                    return std::abs(a - previousY) < std::abs(b - previousY);
                }
            );
            y = *closest;

            const bool isLowerBaselineCandidate =
                panelType == "lower" && y > mask.rows * LOWER_BASELINE_Y_FRACTION;

            if (std::abs(y - previousY) > maxJump && !isLowerBaselineCandidate) {
                previousY = -1;
                continue;
            }
        }

        points.emplace_back(x, y);
        previousY = y;
    }

    return points;
}

InterpolationResult interpolateSmallGaps(const std::vector<cv::Point>& points, int maxGap) {
    InterpolationResult result;

    if (points.size() < 2) {
        result.points = points;
        return result;
    }

    result.points.push_back(points.front());

    for (size_t index = 1; index < points.size(); ++index) {
        const cv::Point previous = result.points.back();
        const cv::Point current = points[index];
        const int gap = current.x - previous.x;

        if (gap > 1 && gap <= maxGap) {
            for (int offset = 1; offset < gap; ++offset) {
                const double ratio = static_cast<double>(offset) / gap;
                const int filledY = static_cast<int>(std::llround(previous.y + (current.y - previous.y) * ratio));
                const cv::Point filled(previous.x + offset, filledY);
                result.points.push_back(filled);
                result.interpolatedPoints.push_back(filled);
            }
        }

        result.points.push_back(current);
    }

    return result;
}

void savePointsCsv(const std::string& outputPath, const std::vector<cv::Point>& points) {
    ensureParentDir(outputPath);
    std::ofstream csvFile(outputPath);
    csvFile << "x_px,y_px\n";
    for (const cv::Point& point : points) {
        csvFile << point.x << "," << point.y << "\n";
    }
}

cv::Mat makeSignalOverlay(
    const cv::Mat& panelImage,
    const cv::Mat& mask,
    const std::vector<cv::Point>& tracePoints,
    const std::vector<cv::Point>& interpolatedPoints,
    const std::string& panelType,
    const std::string& cleanupMode
) {
    cv::Mat overlay = panelImage.clone();
    cv::Mat green(panelImage.size(), panelImage.type(), cv::Scalar(0, 255, 0));
    cv::Mat blended;
    cv::addWeighted(overlay, 0.35, green, 0.65, 0.0, blended);
    blended.copyTo(overlay, mask > 0);

    const int traceGapLimit = panelType == "upper" ? TRACE_GAP_LIMIT_UPPER : TRACE_GAP_LIMIT_LOWER;
    for (size_t index = 1; index < tracePoints.size(); ++index) {
        const cv::Point previous = tracePoints[index - 1];
        const cv::Point current = tracePoints[index];
        if (current.x - previous.x <= traceGapLimit) {
            cv::line(overlay, previous, current, cv::Scalar(0, 0, 255), 2, cv::LINE_AA);
        }
    }

    for (const cv::Point& point : interpolatedPoints) {
        cv::circle(overlay, point, 2, cv::Scalar(0, 255, 255), cv::FILLED, cv::LINE_AA);
    }

    if (panelType == "lower") {
        const int roiYEnd = static_cast<int>(panelImage.rows * getLowerRoiFraction(cleanupMode));
        cv::line(
            overlay,
            cv::Point(0, roiYEnd),
            cv::Point(panelImage.cols - 1, roiYEnd),
            cv::Scalar(255, 0, 0),
            2,
            cv::LINE_AA
        );
        drawLabel(overlay, 14, roiYEnd - 8, "lower ROI", cv::Scalar(255, 0, 0));
    }

    return overlay;
}

int maxInterpolationGapForPanel(const std::string& panelType) {
    return panelType == "upper" ? UPPER_MAX_INTERPOLATION_GAP : LOWER_MAX_INTERPOLATION_GAP;
}

double traceCoveragePercent(const std::vector<cv::Point>& points, int panelWidth) {
    return 100.0 * static_cast<double>(points.size()) / std::max(1, panelWidth);
}

File: python/segmentation.py
import cv2
import numpy as np


MIN_COMPONENT_AREA = 8
MIN_COMPONENT_WIDTH = 2
MIN_COMPONENT_HEIGHT = 2
BORDER_CLEAN_WIDTH = 8
TRACE_GAP_LIMIT_UPPER = 25
TRACE_GAP_LIMIT_LOWER = 50
MAX_EDGE_ARTIFACT_WIDTH = 24
MIN_EDGE_ARTIFACT_HEIGHT_FRACTION = 0.45
TOP_TEXT_ZONE_FRACTION = 0.12
MAX_TOP_TEXT_COMPONENT_AREA = 300
MAX_JUMP_UPPER = 80
MAX_JUMP_LOWER = 100
UPPER_MAX_INTERPOLATION_GAP = 20
LOWER_MAX_INTERPOLATION_GAP = 45
LOWER_SIGNAL_Y_END_FRACTION = 0.93
LOWER_SIGNAL_Y_END_FRACTION_STRICT = 0.86
LOWER_LONG_COMPONENT_Y_FRACTION = 0.70
LOWER_LONG_COMPONENT_ASPECT_RATIO = 12.0
LOWER_LONG_COMPONENT_WIDTH_FRACTION = 0.10
LOWER_LONG_COMPONENT_MAX_HEIGHT = 6
LOWER_RIGHT_TEXT_X_FRACTION = 0.60
LOWER_RIGHT_TEXT_Y_FRACTION = 0.55
LOWER_RIGHT_TEXT_MAX_AREA = 1200
LOWER_RIGHT_TEXT_MAX_HEIGHT = 90
LOWER_RIGHT_TEXT_MAX_WIDTH = 300
LOWER_HORIZONTAL_KERNEL_WIDTH = 40
LOWER_HORIZONTAL_MIN_WIDTH = 100
LOWER_HORIZONTAL_MIN_ASPECT_RATIO = 12.0
LOWER_HORIZONTAL_MAX_HEIGHT = 6
LOWER_BASELINE_KEEP_MIN_HEIGHT = 10
LOWER_BASELINE_KEEP_MIN_AREA = 30
LOWER_BASELINE_Y_FRACTION = 0.70


def split_panels(aligned_image):
    """Делит выровненное изображение на верхнюю и нижнюю панели."""
    height = aligned_image.shape[0]

    upper_start = int(height * 0.05)
    upper_end = int(height * 0.60)
    lower_start = int(height * 0.62)
    lower_end = int(height * 0.98)

    upper_panel = aligned_image[upper_start:upper_end, :].copy()
    lower_panel = aligned_image[lower_start:lower_end, :].copy()

    return upper_panel, lower_panel


def extract_dark_mask(panel_image):
    """Строит простую сырую маску темных пикселей для первичной проверки сигнала."""
    gray = cv2.cvtColor(panel_image, cv2.COLOR_BGR2GRAY)
    blurred = cv2.GaussianBlur(gray, (5, 5), 0)
    _, raw_mask = cv2.threshold(blurred, 160, 255, cv2.THRESH_BINARY_INV)
    return raw_mask


def clean_signal_mask(mask, panel_type, cleanup_mode="soft"):
    """Очищает сырую маску сигнала от мелкого шума и краевых артефактов."""
    clean = mask.copy()
    mask_height, mask_width = clean.shape[:2]
    debug = {}

    # Убираем черные полосы от границ фотографии, они мешают извлечению графика.
    clean[:, :BORDER_CLEAN_WIDTH] = 0
    clean[:, -BORDER_CLEAN_WIDTH:] = 0
    clean[:BORDER_CLEAN_WIDTH, :] = 0
    clean[-BORDER_CLEAN_WIDTH:, :] = 0

    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2, 2))
    clean = cv2.morphologyEx(clean, cv2.MORPH_OPEN, kernel)
    debug["clean_before_roi_cut"] = clean.copy()

    if panel_type == "lower":
        roi_fraction = get_lower_roi_fraction(cleanup_mode)
        roi_y_end = int(mask_height * roi_fraction)
        clean[roi_y_end:, :] = 0

    debug["clean_after_roi_cut"] = clean.copy()
    debug["mask_before_components"] = clean.copy()

    component_count, labels, stats, _ = cv2.connectedComponentsWithStats(clean, 8)
    filtered = np.zeros_like(clean)
    removed_components = np.zeros_like(clean)

    for component_id in range(1, component_count):
        x = stats[component_id, cv2.CC_STAT_LEFT]
        y = stats[component_id, cv2.CC_STAT_TOP]
        width = stats[component_id, cv2.CC_STAT_WIDTH]
        height = stats[component_id, cv2.CC_STAT_HEIGHT]
        area = stats[component_id, cv2.CC_STAT_AREA]

        if should_remove_component(
            panel_type,
            mask_width,
            mask_height,
            x,
            y,
            width,
            height,
            area,
            cleanup_mode,
        ):
            removed_components[labels == component_id] = 255
            continue
        filtered[labels == component_id] = 255

    close_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2, 3))
    filtered = cv2.morphologyEx(filtered, cv2.MORPH_CLOSE, close_kernel)
    filtered, removed_horizontal = remove_lower_horizontal_artifacts(filtered, panel_type, cleanup_mode)
    removed_components = cv2.bitwise_or(removed_components, removed_horizontal)
    filtered, removed_after_close = filter_signal_components(filtered, panel_type, cleanup_mode)
    removed_components = cv2.bitwise_or(removed_components, removed_after_close)
    debug["components_removed"] = removed_components
    debug["components_kept"] = filtered.copy()
    debug["removed_components_debug"] = removed_components
    debug["components_before_count"] = count_components(debug["mask_before_components"])
    debug["components_after_count"] = count_components(filtered)
    debug["components_removed_count"] = max(
        0,
        debug["components_before_count"] - debug["components_after_count"],
    )

    return filtered, debug


def filter_signal_components(mask, panel_type, cleanup_mode="soft"):
    """Удаляет мелкие компоненты и узкие высокие артефакты у правого края."""
    mask_height, mask_width = mask.shape[:2]
    component_count, labels, stats, _ = cv2.connectedComponentsWithStats(mask, 8)
    filtered = np.zeros_like(mask)
    removed = np.zeros_like(mask)

    for component_id in range(1, component_count):
        x = stats[component_id, cv2.CC_STAT_LEFT]
        y = stats[component_id, cv2.CC_STAT_TOP]
        width = stats[component_id, cv2.CC_STAT_WIDTH]
        height = stats[component_id, cv2.CC_STAT_HEIGHT]
        area = stats[component_id, cv2.CC_STAT_AREA]

        if should_remove_component(
            panel_type,
            mask_width,
            mask_height,
            x,
            y,
            width,
            height,
            area,
            cleanup_mode,
        ):
            removed[labels == component_id] = 255
            continue

        filtered[labels == component_id] = 255

    return filtered, removed


def should_remove_component(
    panel_type,
    mask_width,
    mask_height,
    x,
    y,
    width,
    height,
    area,
    cleanup_mode="soft",
):
    """Решает, удалять ли компоненту, не наказывая реальные вертикальные скачки."""
    aspect_ratio = width / max(1, height)

    if area < MIN_COMPONENT_AREA:
        return True
    if width < MIN_COMPONENT_WIDTH or height < MIN_COMPONENT_HEIGHT:
        return True
    if area < 40 and width < 10 and height < 10:
        return True
    if aspect_ratio > 15 and height < 5:
        return True
    if (
        x + width >= mask_width - BORDER_CLEAN_WIDTH
        and width <= MAX_EDGE_ARTIFACT_WIDTH
        and height >= mask_height * MIN_EDGE_ARTIFACT_HEIGHT_FRACTION
    ):
        return True

    if panel_type == "upper":
        return should_remove_upper_component(mask_height, y, area)

    return should_remove_lower_component(
        panel_type,
        mask_width,
        mask_height,
        x,
        y,
        width,
        height,
        area,
        cleanup_mode,
    )


def should_remove_upper_component(mask_height, y, area):
    """Удаляет верхние/нижние подписи верхней панели и короткий шум."""
    if area < 15:
        return True
    if y < mask_height * 0.08 and area < 800:
        return True
    if y > mask_height * 0.92 and area < 800:
        return True

    return False


def should_remove_lower_component(
    panel_type,
    mask_width,
    mask_height,
    x,
    y,
    width,
    height,
    area,
    cleanup_mode="soft",
):
    """Проверяет нижние оси, подписи и служебные горизонтальные элементы."""
    if panel_type != "lower":
        return False

    aspect_ratio = width / max(1, height)
    if area < 20:
        return True

    is_lower_long_line = (
        y > mask_height * LOWER_LONG_COMPONENT_Y_FRACTION
        and aspect_ratio > LOWER_LONG_COMPONENT_ASPECT_RATIO
        and height <= LOWER_LONG_COMPONENT_MAX_HEIGHT
        and width > mask_width * LOWER_LONG_COMPONENT_WIDTH_FRACTION
    )
    is_right_lower_text = (
        x > mask_width * LOWER_RIGHT_TEXT_X_FRACTION
        and y > mask_height * LOWER_RIGHT_TEXT_Y_FRACTION
        and area < LOWER_RIGHT_TEXT_MAX_AREA
        and height < LOWER_RIGHT_TEXT_MAX_HEIGHT
        and width < LOWER_RIGHT_TEXT_MAX_WIDTH
    )

    if is_lower_long_line or is_right_lower_text:
        return True

    is_low_signal_like = (
        y > mask_height * LOWER_BASELINE_Y_FRACTION
        and height >= LOWER_BASELINE_KEEP_MIN_HEIGHT
        and area >= LOWER_BASELINE_KEEP_MIN_AREA
    )
    if cleanup_mode == "soft" and is_low_signal_like:
        return False

    return False


def get_lower_roi_fraction(cleanup_mode):
    """Возвращает границу нижней панели для strict/soft сравнения."""
    if cleanup_mode == "strict":
        return LOWER_SIGNAL_Y_END_FRACTION_STRICT

    return LOWER_SIGNAL_Y_END_FRACTION


def count_components(mask):
    """Считает число связных компонент без фона."""
    component_count, _, _, _ = cv2.connectedComponentsWithStats(mask, 8)
    return max(0, component_count - 1)


def remove_lower_horizontal_artifacts(mask, panel_type, cleanup_mode="soft"):
    """Удаляет нижние почти горизонтальные служебные линии, даже если они склеены со скачком."""
    removed = np.zeros_like(mask)

    if panel_type != "lower":
        return mask, removed

    mask_height = mask.shape[0]
    horizontal_kernel = cv2.getStructuringElement(
        cv2.MORPH_RECT,
        (LOWER_HORIZONTAL_KERNEL_WIDTH, 3),
    )
    horizontal_mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, horizontal_kernel)
    horizontal_mask[: int(mask_height * LOWER_LONG_COMPONENT_Y_FRACTION), :] = 0

    component_count, labels, stats, _ = cv2.connectedComponentsWithStats(horizontal_mask, 8)

    for component_id in range(1, component_count):
        width = stats[component_id, cv2.CC_STAT_WIDTH]
        height = stats[component_id, cv2.CC_STAT_HEIGHT]
        aspect_ratio = width / max(1, height)

        if (
            width >= LOWER_HORIZONTAL_MIN_WIDTH
            and aspect_ratio >= LOWER_HORIZONTAL_MIN_ASPECT_RATIO
            and height <= LOWER_HORIZONTAL_MAX_HEIGHT
        ):
            removed[labels == component_id] = 255

    cleaned = mask.copy()
    cleaned[removed > 0] = 0

    return cleaned, removed


def extract_trace_points(mask, panel_type):
    """Получает первичную линию графика: по одному y для каждого столбца с сигналом."""
    points = []
    previous_y = None
    max_jump = MAX_JUMP_UPPER if panel_type == "upper" else MAX_JUMP_LOWER

    for x in range(mask.shape[1]):
        ys = np.where(mask[:, x] > 0)[0]

        if len(ys) == 0:
            previous_y = None
            continue

        if previous_y is None:
            y = int(np.median(ys))
        else:
            closest_index = int(np.argmin(np.abs(ys - previous_y)))
            y = int(ys[closest_index])

            is_lower_baseline_candidate = (
                panel_type == "lower"
                and y > mask.shape[0] * LOWER_BASELINE_Y_FRACTION
            )

            if abs(y - previous_y) > max_jump and not is_lower_baseline_candidate:
                previous_y = None
                continue

        points.append((x, y))
        previous_y = y

    return points


def interpolate_small_gaps(points, max_gap):
    """Заполняет короткие разрывы линейной интерполяцией."""
    if len(points) < 2:
        return points, set()

    filled_points = [points[0]]
    interpolated_points = set()

    for point in points[1:]:
        previous_x, previous_y = filled_points[-1]
        x, y = point
        gap = x - previous_x

        if 1 < gap <= max_gap:
            for offset in range(1, gap):
                ratio = offset / gap
                filled_y = int(round(previous_y + (y - previous_y) * ratio))
                filled_point = (previous_x + offset, filled_y)
                filled_points.append(filled_point)
                interpolated_points.add(filled_point)

        filled_points.append(point)

    return filled_points, interpolated_points


def count_large_gaps(points, gap_limit):
    """Считает разрывы по x, которые не надо соединять на overlay."""
    gap_count = 0

    for index in range(1, len(points)):
        if points[index][0] - points[index - 1][0] > gap_limit:
            gap_count += 1

    return gap_count


def make_overlay(
    panel_image,
    mask,
    trace_points=None,
    interpolated_points=None,
    panel_type="upper",
    cleanup_mode="soft",
):
    """Накладывает маску и найденную линию на панель для отладки этапа 2."""
    overlay = panel_image.copy()

    green = np.zeros_like(overlay)
    green[:, :] = (0, 255, 0)
    overlay = np.where(mask[:, :, None] > 0, cv2.addWeighted(overlay, 0.35, green, 0.65, 0), overlay)

    if trace_points:
        previous = None
        trace_gap_limit = TRACE_GAP_LIMIT_UPPER if panel_type == "upper" else TRACE_GAP_LIMIT_LOWER

        for point in trace_points:
            if previous is not None and point[0] - previous[0] <= trace_gap_limit:
                cv2.line(overlay, previous, point, (0, 0, 255), 2)
            previous = point

    if interpolated_points:
        for point in interpolated_points:
            cv2.circle(overlay, point, 2, (0, 255, 255), -1)

    if panel_type == "lower":
        roi_y_end = int(panel_image.shape[0] * get_lower_roi_fraction(cleanup_mode))
        cv2.line(overlay, (0, roi_y_end), (panel_image.shape[1] - 1, roi_y_end), (255, 0, 0), 2)

    return overlay


def make_trace_only(
    panel_shape,
    trace_points=None,
    interpolated_points=None,
    panel_type="upper",
    cleanup_mode="soft",
):
    """Создает отдельное изображение только с трассой и интерполяцией."""
    height, width = panel_shape[:2]
    trace_image = np.zeros((height, width, 3), dtype=np.uint8)

    if trace_points:
        previous = None
        trace_gap_limit = TRACE_GAP_LIMIT_UPPER if panel_type == "upper" else TRACE_GAP_LIMIT_LOWER

        for point in trace_points:
            if previous is not None and point[0] - previous[0] <= trace_gap_limit:
                cv2.line(trace_image, previous, point, (0, 0, 255), 2)
            previous = point

    if interpolated_points:
        for point in interpolated_points:
            cv2.circle(trace_image, point, 2, (0, 255, 255), -1)

    if panel_type == "lower":
        roi_y_end = int(height * get_lower_roi_fraction(cleanup_mode))
        cv2.line(trace_image, (0, roi_y_end), (width - 1, roi_y_end), (255, 0, 0), 2)

    return trace_image


def analyze_panel(panel_image, panel_type):
    """Выполняет стартовую сегментацию одной панели."""
    raw_mask = extract_dark_mask(panel_image)
    clean_mask, debug = analyze_clean_mask(panel_image, raw_mask, panel_type, "soft")

    if panel_type == "lower":
        strict_clean_mask, strict_debug = analyze_clean_mask(panel_image, raw_mask, panel_type, "strict")
        before_cleanup_points = extract_trace_points(strict_debug["clean_before_roi_cut"], panel_type)
        debug["strict_clean_mask"] = strict_clean_mask
        debug["strict_overlay"] = strict_debug["overlay"]
        debug["strict_debug"] = strict_debug
        debug["trace_points_before_cleanup"] = len(before_cleanup_points)
        debug["trace_points_after_cleanup"] = len(debug["interpolated_trace_points"])
        debug["selected_mode"] = "soft"

    return raw_mask, clean_mask, debug["interpolated_trace_points"], debug["overlay"], debug


def analyze_clean_mask(panel_image, raw_mask, panel_type, cleanup_mode):
    """Строит clean-маску и трассу для выбранного режима очистки."""
    clean_mask, debug = clean_signal_mask(raw_mask, panel_type, cleanup_mode)
    raw_trace_points = extract_trace_points(clean_mask, panel_type)
    max_gap = UPPER_MAX_INTERPOLATION_GAP if panel_type == "upper" else LOWER_MAX_INTERPOLATION_GAP
    trace_points, interpolated_points = interpolate_small_gaps(raw_trace_points, max_gap)
    overlay = make_overlay(
        panel_image,
        clean_mask,
        trace_points,
        interpolated_points,
        panel_type,
        cleanup_mode,
    )
    trace_only = make_trace_only(
        panel_image.shape,
        trace_points,
        interpolated_points,
        panel_type,
        cleanup_mode,
    )

    trace_gap_limit = TRACE_GAP_LIMIT_UPPER if panel_type == "upper" else TRACE_GAP_LIMIT_LOWER
    debug["raw_trace_points"] = raw_trace_points
    debug["interpolated_trace_points"] = trace_points
    debug["interpolated_points_set"] = interpolated_points
    debug["overlay"] = overlay
    debug["trace_only"] = trace_only
    debug["raw_points_count"] = len(raw_trace_points)
    debug["interpolated_points_count"] = len(trace_points)
    debug["large_gaps_count"] = count_large_gaps(raw_trace_points, trace_gap_limit)
    debug["trace_coverage"] = 100.0 * len(trace_points) / max(1, panel_image.shape[1])
    debug["cleanup_mode"] = cleanup_mode

    return clean_mask, debug
